File: problems/1089-duplicate-zeros/solution.c
void duplicateZeros(int* arr, int arrSize) {
    int zeros = 0;
    for (int i = 0; i < arrSize; i++) {
        if (arr[i] == 0) {
            zeros++;
        }
    }

    int i = arrSize - 1;
    int j = arrSize + zeros - 1;

    while (i != j) {
        if (arr[i] == 0) {
            if (j < arrSize) {
                arr[j] = 0;
            }
            j--;
            if (j < arrSize) {
                arr[j] = 0;
            }
        } else {
            if (j < arrSize) {
                arr[j] = arr[i];
            }
        }
        i--;
        j--;
    }
}

File: problems/1089-duplicate-zeros/solution.java
class Solution {
    public void duplicateZeros(int[] arr) {
        int zeros = 0;
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == 0) {
                zeros++;
            }
        }

        int n = arr.length;
        for (int i = n - 1; i >= 0; i--) {
            if (arr[i] == 0) {
                if (i + zeros < n) {
                    arr[i + zeros] = 0;
                }
                zeros--;
                if (i + zeros < n) {
                    arr[i + zeros] = 0;
                }
            } else {
                if (i + zeros < n) {
                    arr[i + zeros] = arr[i];
                }
            }
        }
    }
}